File: lista/Lista Estática/Java/Lista Circular/ListaEstaticaCircular.java
public class ListaEstaticaCircular implements Listavel {
	
	private int ponteiroInicio; 
	private int ponteiroFim;
	private Object[] dados;
	private int quantidade;
	
	public ListaEstaticaCircular() {
		this(10);
	}
	
	public ListaEstaticaCircular(int tamanho) {
		ponteiroInicio = 0;
		ponteiroFim = -1;
		quantidade = 0;		
		dados = new Object[tamanho];
	}

	//DE endereçamento lógico (informado pelo usuário)
	//PARA endereçamento físico (onde o dado está no array)
	private int mapeamento(int logica) {
		return (logica + ponteiroInicio)%dados.length;
	}
	
	private int avancar(int ponteiro) {
		return (ponteiro+1)%dados.length;
	}
	
	private int retroceder(int ponteiro) {
		return ((ponteiro-1)+dados.length)%dados.length;
	}		

	//funciona como o estaCheia de FilaEstaticaCircular
	@Override
	public boolean estaCheia() {
		return (quantidade == dados.length);
	}
	
	//funciona como o estaVazia de FilaEstaticaCircular
	@Override
	public boolean estaVazia() {
		return (quantidade == 0);
	}
	
	//funciona como o imprimir de FilaEstaticaCircular
	@Override
	public String imprimir(){
		String retorno = "[";
		int ponteiroAux = ponteiroInicio;
		for (int i = 0; i < quantidade; i++) {			
			retorno += dados[ponteiroAux];
			if (i != quantidade - 1) 
				retorno += ",";
			
			ponteiroAux = avancar(ponteiroAux); 
		}
		return retorno + "]";
	}
	
	//funciona como o enfileirar de FilaEstaticaCircular
	@Override
	public void anexar(Object dado) {
		if (!estaCheia()) {
			ponteiroFim = avancar(ponteiroFim);
			dados[ponteiroFim] = dado;			
			quantidade++;
		} else {
			System.err.println("Lista Cheia!");
		}			
	}

	@Override
	public Object[] selecionarTodos() {
		Object[] dadosAux = null;
		if (!estaVazia()) {
			dadosAux = new Object[quantidade];
			int ponteiroAux = ponteiroInicio;
			for(int i = 0; i < quantidade; i++) {
				dadosAux[i] = dados[ponteiroAux];
				ponteiroAux = avancar(ponteiroAux);
			}
		} else {
			System.err.println("List is empty!");
		}
		return dadosAux;
	}

	@Override
	public Object selecionar (int posicao) {
		Object dadoAux = null;
		if (!estaVazia()) {
			//verificando se o índice/posição é válido 
			if ((posicao >= 0) && (posicao < quantidade)) {				
				int posicaoFisica = mapeamento(posicao);
				dadoAux = dados[posicaoFisica];
			} else {
				System.err.println("Indice Invalido!");	
			}
		} else {
			System.err.println("Lista Vazia!");		
		}
		return dadoAux;
	}

	@Override
	public void atualizar(int posicao, Object novoDado) {
		if (!estaVazia()) {
			//verificando se o índice/posição é válido
			if ((posicao >= 0) && (posicao < quantidade)) {
				int posicaoFisica = mapeamento(posicao);
				dados[posicaoFisica] = novoDado;
			} else {
				System.err.println("Indice Invalido!");	
			}
		} else {
			System.err.println("Lista Vazia!");
		}
	}

	@Override
	public void inserir(int posicao, Object dado) {
		if (!estaCheia()) {
			if (posicao >= 0 && posicao <= quantidade) {
				int posicaoFisica = mapeamento(posicao);
				int x = ponteiroFim;
				int y = avancar(x);
				for (int i = 0; i < quantidade-posicao;i++) {
					dados[x] = dados[y];
					x = retroceder(x);
					y = retroceder(y);
				}
				dados[posicaoFisica] = dado;
				quantidade++;
				avancar(ponteiroFim);
			} else {
				System.err.println("Invalid Index!");
			}
		} else {
			System.err.println("List is empty!");
		}
	}
	
	@Override
	public Object apagar(int posicao) {
		Object dadoAux = null;
		if (!estaVazia()) {
			if (posicao >= 0 && posicao < quantidade) {
				int posicaoFisica = mapeamento(posicao);
				dadoAux = dados[posicaoFisica];
				int x = posicaoFisica;
				int y = avancar(x);
				for (int i = 0; i < quantidade-posicao-1;i++) {	
					dados[x] = dados[y];
					x = avancar(x);
					y = avancar(y);					
				}
				quantidade--;
				ponteiroFim = retroceder(ponteiroFim);
			} else {
				System.err.println("Invalid Index!");
			}
		} else {
			System.err.println("List is empty!");
		}
		return dadoAux;
	}
}

File: lista/Lista Estática/Kotlin/Lista Circular/ListaEstaticaCircular.kt
class ListaEstaticaCircular : Listavel {
 
	private var ponteiroInicio: Int
	private var ponteiroFim: Int
	private var dados: Array<Any?>
	private var quantidade: Int

	constructor(tamanho: Int) {
		dados = arrayOfNulls(tamanho)
		ponteiroInicio = 0
		ponteiroFim = -1
		quantidade = 0	
	}
	
	constructor() : this(10) {}
    
	override fun anexar(dado: Any?) {
		if (!estaCheia()) {
			ponteiroFim++
			if (ponteiroFim == dados.size)
				ponteiroFim = 0

			quantidade++			
			dados[ponteiroFim] = dado
		} else {
			println("List is full!")
		}
	}

	override fun selecionarTodos(): Array<Any?> {        
		var dadosAux: Array<Any?> = arrayOfNulls(quantidade)
		if (!estaVazia()) {
			var ponteiroAux: Int = ponteiroInicio
            for (i in 0 until quantidade)  {
				if (ponteiroAux == dados.size) {
                    ponteiroAux = 0
                }
				dadosAux[i] = dados[ponteiroAux]
				ponteiroAux++
            }
        }
        return dadosAux
	}

	override fun selecionar (posicao: Int): Any? {
		var dadoAux: Any? = null
		if (!estaVazia()) {
			//índice/posição é válido? 
			if (posicao >= 0 && 
					posicao < quantidade) {
				//mapeamento:
				//DE endereçamento lógico 
				//(informado pelo usuário)
				//PARA endereçamento físico 
				//(onde o dado está no array)
				var posicaoFisica: Int = 
					(ponteiroInicio + posicao) % dados.size
				dadoAux = dados[posicaoFisica]
			} else {
				println("Indice Invalido!")	
			}
		} else {
			println("Lista Vazia!")		
		}
		return dadoAux
	}

	override fun atualizar (posicao: Int, dado: Any?) {
		if (!estaVazia()) {
			//índice/posição é válido? 
			if (posicao >= 0 && 
					posicao < quantidade) {
				//mapeamento:
				//DE endereçamento lógico 
				//(informado pelo usuário)
				//PARA endereçamento físico 
				//(onde o dado está no array)
				var posicaoFisica: Int = 
					(ponteiroInicio + posicao) % dados.size
				dados[posicaoFisica] = dado
			} else {
				println("Indice Invalido!")	
			}
		} else {
			println("Lista Vazia!")		
		}
	}

	override fun estaCheia(): Boolean {
		return quantidade == dados.size
	}
	
	override fun estaVazia(): Boolean {
		return quantidade == 0
	}
	
	override fun imprimir(): String {
		var resultado : String = "["
		var ponteiroAux: Int = ponteiroInicio
		for (i in 0 until quantidade) {
			if (i == ponteiroFim)
				resultado += 
					"${dados[ponteiroAux % dados.size]}"
			else
				resultado += 
					"${dados[ponteiroAux % dados.size]}, "
			
			ponteiroAux++				
		}
		return resultado + "]"
	}

	override fun apagar(posicao: Int): Any? {
		var dadoAux: Any? = null
		if (!estaVazia()) {
			//índice/posição é válido? 
			if (posicao >= 0 && 
					posicao < quantidade) {
				//mapeamento:
				//DE endereçamento lógico 
				//(informado pelo usuário)
				//PARA endereçamento físico 
				//(onde o dado está no array)
				var posicaoFisica: Int = 
				(ponteiroInicio + posicao) % dados.size
				dadoAux = dados[posicaoFisica]
				var ponteiroAux: Int = posicaoFisica
				for (i in posicao until (quantidade-1)) {
					var atual: Int = ponteiroAux
					if (ponteiroAux == dados.size-1) {
						ponteiroAux = -1			
					}
					var proximo: Int = ponteiroAux+1

					dados[atual] = dados[proximo]
					ponteiroAux++
				}
				ponteiroFim--
				if (ponteiroFim == -1) {
					ponteiroFim = dados.size - 1
				}
				quantidade--
			} else {
				println("Indice Invalido!")
			}
		} else {
			println("Lista Vazia!")
		}
		return dadoAux
	}

	override fun inserir(posicao: Int, dado: Any?) {
		if (!estaCheia()) {
			//índice/posição é válido? 
			if (posicao >= 0 && 
					posicao <= quantidade) {
				//mapeamento:
				//DE endereçamento lógico 
				//(informado pelo usuário)
				//PARA endereçamento físico 
				//(onde o dado está no array)
				var posicaoFisica: Int = 
					(ponteiroInicio + posicao) % dados.size

				var ponteiroAux: Int = ponteiroFim+1
				for (i in posicao until quantidade) {
					var anterior: Int = ponteiroAux-1

					if(ponteiroAux == dados.size) {
						ponteiroAux = 0			
					} 
					var atual: Int = ponteiroAux

					dados[atual] = dados[anterior]
					ponteiroAux--
				}

				dados[posicaoFisica] = dado
				ponteiroFim++
				if (ponteiroFim == dados.size) {
					ponteiroFim = 0
				}
				quantidade++
			} else {
				println("Indice Invalido")
			}
		} else {
			println("Lista Cheia!")
		}
	}
}